src/core/sock/sock-extra.cpp

```diff
@@ -373,3 +375,189 @@ struct xlio_api_t *extra_api(void)
+{
+    // Set XLIO socket API specific parameter unless user sets them explicitly
+    if (!getenv(SYS_VAR_PROGRESS_ENGINE_INTERVAL)) {
+        setenv(SYS_VAR_PROGRESS_ENGINE_INTERVAL, "0", 1);
+    }
+    if (!getenv(SYS_VAR_TCP_ABORT_ON_CLOSE)) {
+        setenv(SYS_VAR_TCP_ABORT_ON_CLOSE, "1", 1);
+    }
+
+    extern xlio_memory_cb_t g_user_memory_cb;
+    g_user_memory_cb = attr->memory_cb;
+
+    xlio_init();
+    DO_GLOBAL_CTORS();
+
+    return 0;
+}
+
+extern "C" int xlio_group_create(const struct xlio_group_attr *attr, xlio_group_t *group_out)
+{
+    // Validate input arguments
+    if (!group_out || !attr || !attr->socket_event_cb) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    xlio_group *grp = new xlio_group(attr);
+    if (!grp) {
+        errno = ENOMEM;
+        return -1;
+    }
+
+    *group_out = reinterpret_cast<xlio_group_t>(grp);
+    return 0;
+}
+
+extern "C" int xlio_group_destroy(xlio_group_t group)
+{
+    xlio_group *grp = reinterpret_cast<xlio_group *>(group);
+
+    delete grp;
+    return 0;
+}
+
+extern "C" void xlio_group_poll(xlio_group_t group)
+{
+    xlio_group *grp = reinterpret_cast<xlio_group *>(group);
+
+    grp->poll();
+}
+
+extern "C" int xlio_socket_create(const struct xlio_socket_attr *attr, xlio_socket_t *sock_out)
+{
+    // Validate input arguments
+    if (!sock_out || !attr || !attr->group ||
+        !(attr->domain == AF_INET || attr->domain == AF_INET6)) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    int sockfd = socket_internal(attr->domain, SOCK_STREAM, 0, true, false);
+    if (sockfd < 0) {
+        return -1;
+    }
+
+    sockinfo_tcp *si = dynamic_cast<sockinfo_tcp *>(g_p_fd_collection->get_sockfd(sockfd));
+    if (!si) {
+        errno = EBADF;
+        return -1;
+    }
+
+    si->set_xlio_socket(attr);
+
+    *sock_out = reinterpret_cast<xlio_socket_t>(si);
+    return 0;
+}
+
+extern "C" int xlio_socket_destroy(xlio_socket_t sock)
+{
+    sockinfo_tcp *si = reinterpret_cast<sockinfo_tcp *>(sock);
+
+    return XLIO_CALL(close, si->get_fd());
+}
+
+extern "C" int xlio_socket_setsockopt(xlio_socket_t sock, int level, int optname,
+                                      const void *optval, socklen_t optlen)
+{
+    sockinfo_tcp *si = reinterpret_cast<sockinfo_tcp *>(sock);
+
+    return XLIO_CALL(setsockopt, si->get_fd(), level, optname, optval, optlen);
+}
+
+extern "C" int xlio_socket_bind(xlio_socket_t sock, const struct sockaddr *addr, socklen_t addrlen)
+{
+    sockinfo_tcp *si = reinterpret_cast<sockinfo_tcp *>(sock);
+
+    return XLIO_CALL(bind, si->get_fd(), addr, addrlen);
+}
+
+extern "C" int xlio_socket_connect(xlio_socket_t sock, const struct sockaddr *to, socklen_t tolen)
+{
+    sockinfo_tcp *si = reinterpret_cast<sockinfo_tcp *>(sock);
+
+    int rc = XLIO_CALL(connect, si->get_fd(), to, tolen);
+    rc = (rc == -1 && (errno == EINPROGRESS || errno == EAGAIN)) ? 0 : rc;
+    if (rc == 0) {
+        si->add_tx_ring_to_group();
```

Comment:
What happens with non-blocking socket? when will it be added to the group?

Reply:
Done, errno is restored for EINPROGRESS case (as it's a successful call).